File: Python/1518. Water Bottles.py
class Solution:
    def numWaterBottles(self, numBottles: int, numExchange: int) -> int:
        def exchange(fullbottles, empty, total):
            if (fullbottles + empty < numExchange):
                return total + fullbottles
            return exchange(empty // numExchange, fullbottles + empty - empty // numExchange * numExchange, total + fullbottles)

        return exchange(numBottles, 0, 0)

File: Javascript/1518. Water Bottles.js
/**
 * @param {number} numBottles
 * @param {number} numExchange
 * @return {number}
 */
var numWaterBottles = function (numBottles, numExchange) {
  let total = 0;

  const drink = (bottle, empty) => {
    if (empty < numExchange && bottle === 0) return;
    total += bottle;
    const emp = bottle + empty;
    const curr = Math.floor(emp / numExchange);
    drink(curr, emp - curr * numExchange);
  };

  drink(numBottles, 0);
  return total;
};
